src/WebJobs.Script/Workers/Rpc/Configuration/DynamicWorkerConfigurationResolver.cs

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using System.Text.Json;
+using Microsoft.Azure.WebJobs.Script.Workers.Profiles;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.Workers.Rpc.Configuration
+{
+    // This class resolves worker configurations dynamically based on the current environment and configuration settings.
+    // It searches for worker configs in specified probing paths and the fallback path, and returns a list of worker configuration paths.
+    internal sealed class DynamicWorkerConfigurationResolver : IWorkerConfigurationResolver
+    {
+        private readonly ILogger _logger;
+        private readonly IWorkerProfileManager _profileManager;
+        private readonly IFileSystem _fileSystem;
+        private readonly HashSet<string> _workersAvailableForResolution;
+        private readonly List<string> _workerProbingPaths;
+        private readonly JsonSerializerOptions _jsonSerializerOptions = new() { PropertyNameCaseInsensitive = true };
+        private readonly IOptionsMonitor<WorkerConfigurationResolverOptions> _workerConfigurationResolverOptions;
+        private readonly Dictionary<string, HashSet<Version>> _ignoredVersions;
+
+        public DynamicWorkerConfigurationResolver(ILoggerFactory loggerFactory,
+                                                    IFileSystem fileSystem,
+                                                    IWorkerProfileManager workerProfileManager,
+                                                    IOptionsMonitor<WorkerConfigurationResolverOptions> workerConfigResolverOptions)
+        {
+            ArgumentNullException.ThrowIfNull(loggerFactory);
+            _logger = loggerFactory.CreateLogger(ScriptConstants.LogCategoryWorkerConfig);
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _profileManager = workerProfileManager ?? throw new ArgumentNullException(nameof(workerProfileManager));
+            _workerConfigurationResolverOptions = workerConfigResolverOptions ?? throw new ArgumentNullException(nameof(workerConfigResolverOptions));
+            ArgumentNullException.ThrowIfNull(workerConfigResolverOptions.CurrentValue);
+            _workerProbingPaths = workerConfigResolverOptions.CurrentValue.ProbingPaths;
+            _workersAvailableForResolution = workerConfigResolverOptions.CurrentValue.WorkersAvailableForResolution;
+            _ignoredVersions = workerConfigResolverOptions.CurrentValue.IgnoredWorkerVersions;
```

Comment:
You can potentially use the _workerConfigurationResolverOptions instead of `_workerProbingPaths ,_workersAvailableForResolution, _ignoredVersions)`  The properties here are access only once throughout the class. They likely don't need to cached in the constructor.